action/jsonapi.go

```diff
@@ -0,0 +1,210 @@
+package action
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"github.com/justwatchcom/gopass/store"
+	"github.com/urfave/cli"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type messageType struct {
+	Type string `json:"type"`
+}
+
+type queryMessage struct {
+	Query string `json:"query"`
+}
+
+type getLoginMessage struct {
+	Entry string `json:"entry"`
+}
+
+type loginResponse struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type errorResponse struct {
+	Error string `json:"error"`
+}
+
+// JSONAPI reads a json message on stdin and responds on stdout
+func (s *Action) JSONAPI(ctx context.Context, c *cli.Context) error {
+	err := readAndRespond(ctx, s)
+	if err != nil {
+		var response errorResponse
+		response.Error = err.Error()
+
+		err = sendSerializedJSONMessage(response)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func readAndRespond(ctx context.Context, s *Action) error {
+	message, err := readMessage()
+	if message != nil {
```

Comment:
I'd prefer to handle the error case in the block and the success case on the upper level, e.g. `if err != nil || message == nil { <error handling>  } <success handling>`